release/release.go

```diff
@@ -81,51 +128,107 @@ func GenReleaseNotes(ctx context.Context, repo, milestone, prevMilestone string,
 	majorMinor := tmp[0] + "." + tmp[1]
-	fullVersion := majorMinor + "." + tmp[2]
 	changeLogSince := strings.Replace(strings.Split(prevMilestone, "+")[0], ".", "", -1)
 	sqliteVersionK3S := goModLibVersion("go-sqlite3", repo, milestone)
 	sqliteVersionBinding := sqliteVersionBinding(sqliteVersionK3S)
+	helmControllerVersion := goModLibVersion("helm-controller", repo, milestone)
+	coreDNSVersion := imageTagVersion("coredns", repo, milestone)
+	cgData := changeLogData{
+		PrevMilestone: prevMilestone,
+		Content:       content,
+	}
+
+	if repo == k3sRepo {
+		return genK3SReleaseNotes(
+			tmpl,
+			milestone,
+			k3sReleaseNoteData{
+				Milestone:             milestoneNoRC,
+				MajorMinor:            majorMinor,
+				K8sVersion:            k8sVersion,
+				ChangeLogVersion:      markdownVersion,
+				ChangeLogSince:        changeLogSince,
+				SQLiteVersion:         sqliteVersionBinding,
+				SQLiteVersionReplaced: strings.ReplaceAll(sqliteVersionBinding, ".", "_"),
+				HelmControllerVersion: helmControllerVersion,
+				ChangeLogData:         cgData,
+				CoreDNSVersion:        coreDNSVersion,
+			},
+		)
+	} else if repo == rke2Repo {
+		return genRKE2ReleaseNotes(
+			tmpl,
+			milestone,
+			rke2ReleaseNoteData{
+				MajorMinor:            majorMinor,
+				Milestone:             milestoneNoRC,
+				ChangeLogVersion:      markdownVersion,
+				K8sVersion:            k8sVersion,
+				HelmControllerVersion: helmControllerVersion,
+				CoreDNSVersion:        coreDNSVersion,
+				ChangeLogData:         cgData,
+			},
+		)
+	}
+	return nil, errors.New("invalid repo, it must be either k3s or rke2")
+}
+
+func genK3SReleaseNotes(tmpl *template.Template, milestone string, rd k3sReleaseNoteData) (*bytes.Buffer, error) {
+	tmpl = template.Must(tmpl.Parse(k3sReleaseNoteTemplate))
+	var runcVersion string
+	var containerdVersion string
+
+	if semver.Compare(rd.K8sVersion, "v1.24.0") == 1 &&
+		semver.Compare(rd.K8sVersion, "v1.26.5") == -1 {
+		containerdVersion = buildScriptVersion("VERSION_CONTAINERD", k3sRepo, milestone)
+	} else {
+		containerdVersion = goModLibVersion("containerd/containerd", k3sRepo, milestone)
+	}
+
+	if rd.MajorMinor == "1.23" {
+		runcVersion = buildScriptVersion("VERSION_RUNC", k3sRepo, milestone)
+	} else {
+		runcVersion = goModLibVersion("runc", k3sRepo, milestone)
+	}
+
+	rd.KineVersion = goModLibVersion("kine", k3sRepo, milestone)
+	rd.EtcdVersion = goModLibVersion("etcd/api/v3", k3sRepo, milestone)
+	rd.ContainerdVersion = containerdVersion
+	rd.RuncVersion = runcVersion
+	rd.FlannelVersion = goModLibVersion("flannel", k3sRepo, milestone)
+	rd.MetricsServerVersion = imageTagVersion("metrics-server", k3sRepo, milestone)
+	rd.TraefikVersion = imageTagVersion("traefik", k3sRepo, milestone)
+	rd.LocalPathProvisionerVersion = imageTagVersion("local-path-provisioner", k3sRepo, milestone)
 
 	buf := bytes.NewBuffer(nil)
+	err := tmpl.ExecuteTemplate(buf, k3sRepo, rd)
+	return buf, err
```

Comment:
I think we should return nil here rather than an initialized pointer for no reason.
```suggestion
	return nil, err
```